### PR TITLE
ContainerBase: Improve ports things

### PR DESCRIPTION
**Pull request by JayH5**
* get_host_port default to TCP ports
* get_host_port returns a tuple of IP address, port
* ports property returns all port mappings
* get_first_host_port for easy host port lookup when there is only one port
* Much more error checking for when ports not exposed or published

### CODE REVIEW

File: seaworthy/containers/base.py [hunk 1]
```diff
@@ -115,15 +115,54 @@ def clean(self):
...
+    def get_first_host_port(self):
+        """
+        Get the first mapping of the first (lowest) container port that has a
+        mapping. Useful when a container publishes only one port.
+
+        Note that unlike the Docker API, which sorts ports alphabetically (e.g.
+        ``90/tcp`` > ``8000/tcp``), we sort ports numerically so that the
+        lowest port is always chosen.
+        """
+        mapped_ports = {p: m for p, m in self.ports.items() if m is not None}
```
Comment (JayH5) [inline seaworthy/containers/base.py]: Not sure if it's worth having a `published_ports` or `mapped_ports` property?

File: seaworthy/containers/base.py [hunk 2]
```diff
@@ -115,15 +115,54 @@ def clean(self):
...
+        if port_spec not in self.ports:
+            raise ValueError("Port '{}' is not exposed".format(port_spec))
+
+        mappings = self.ports[port_spec]
+        if mappings is None:
+            raise ValueError(
+                "Port '{}' is not published to the host".format(port_spec))
+
+        mapping = mappings[index]
+        return mapping['HostIp'], mapping['HostPort']
```
Comment (JayH5) [inline seaworthy/containers/base.py]: Maybe it'd be nice to turn the IP into an [`ipaddress`](https://docs.python.org/3/library/ipaddress.html)?

File: seaworthy/containers/base.py [hunk 3]
```diff
@@ -115,15 +115,54 @@ def clean(self):
...
-        return ports[port_spec][index]['HostPort']
+        port_spec = '{}/{}'.format(container_port, proto)
+        return self._host_port(port_spec, index)
+
+    def get_first_host_port(self):
+        """
+        Get the first mapping of the first (lowest) container port that has a
+        mapping. Useful when a container publishes only one port.
+
+        Note that unlike the Docker API, which sorts ports alphabetically (e.g.
```
Comment (jerith) [inline seaworthy/containers/base.py]: I think `lexicographically` is more correct than `alphabetically`.

File: seaworthy/containers/base.py (hunk 1 above)
Comment (jerith) [reply seaworthy/containers/base.py]: Not unless you have a concrete use case, I think.

File: seaworthy/containers/base.py (hunk 2 above)
Comment (jerith) [reply seaworthy/containers/base.py]: Unless we're happy that it won't break all the things expecting a string (like wedging it into a URL or whatever) it's probably best to leave it for now.

**Pull request merged**